core/src/main/java/org/jdbi/v3/BuiltInArgumentFactory.java

```diff
@@ -90,30 +123,55 @@ public Argument build(Class<?> expectedType, T value, StatementContext ctx)
 
         // Enums must be bound as VARCHAR.
         if (expectedType.isEnum()) {
-            return new StringArgument(value.toString());
+            return STR_BUILDER.build(value.toString());
         }
 
         // Fallback to generic ObjectArgument
-        return new ObjectArgument(value);
+        return OBJ_BUILDER.build(value);
     }
 
-    private static class P
-    {
-        private final Constructor<?> ctor;
+    @FunctionalInterface
+    interface StatementBinder<T> {
+        void bind(PreparedStatement p, int index, T value) throws SQLException;
+    }
```

Comment:
:+1: Elegant solution, taking advantage of the uniform method signatures in ResultSet.